packages/flutter/lib/src/material/segmented_button.dart

```diff
@@ -0,0 +1,770 @@
+import 'icons.dart';
+import 'material.dart';
+import 'material_state.dart';
+import 'segmented_button_theme.dart';
+import 'text_button.dart';
+import 'text_button_theme.dart';
+import 'theme.dart';
+
+/// Data describing a segment of a [SegmentedButton].
+class ButtonSegment<T> {
+  /// Construct a SegmentData
+  ///
+  /// One of [icon] or [label] must be non-null.
+  const ButtonSegment({
+    required this.value,
+    this.icon,
+    this.label,
+    this.enabled = true,
+  }) : assert(icon != null || label != null);
+
+  /// Value used to identify the segment.
+  ///
+  /// This value must be unique across all segments in a [SegmentedButton].
+  final T value;
+
+  /// Optional icon displayed in the segment.
+  final Widget? icon;
+
+  /// Optional label displayed in the segment.
+  final Widget? label;
+
+  /// Determines if the segment is available for selection.
+  final bool enabled;
+}
+
+/// A Material button that allows the user to select from limited set of options.
+///
+/// Segmented buttons are used to help people select options, switch views, or
+/// sort elements. They are typically used in cases where there are only 2-5
+/// options.
+///
+/// The options are represented by segments described with [ButtonSegment]
+/// entries in the [segments] field. Each segment has a [ButtonSegment.value]
+/// that is used to indicated which segment's are selected.
+///
+/// The [selected] field is a set of the selected segments' values. This should
+/// be updated by the app in response to [onSelectionChanged] updates.
+///
+/// By default only a single segment can be selected (for mutually exclusive
+/// choices). This can be relaxed with the [multiSelectionEnabled] field.
+///
+/// Like [ButtonStyleButton]s the [SegmentedButton]'s visuals can be
+/// configured with a [ButtonStyle] [style] field. However, unlike other
+/// buttons, some of the style parameters are applied to the entire segmented
+/// button, and others are used for each of the segments.
+///
+/// By default a checkmark icon is used to show selected items. To configure
+/// this behavior, you can use the [showSelectedIcon] and [selectedIcon] fields.
+///
+/// Individual segments can be enabled or disabled with their
+/// [ButtonSegment.enabled] flag, but if [onSelectionChanged] field is null
+/// then the entire segmented button will be disabled, regardless of the
+/// individual segment settings.
+///
+/// See also:
+///
+///   * Material Design spec: <https://m3.material.io/components/segmented-buttons/overview>
+///   * [ButtonStyle], which can be used to configure the appearance of the button.
+///   * [Radio], an alternative way to present the user with a mutually exclusive set of options.
+///   * [FilterChip], [ChoiceChip], which can be used when you need to show more then five options.
+class SegmentedButton<T> extends StatelessWidget {
+  /// Creates a segmented button.
+  ///
+  /// [segments] must contain at least one segment, but it is recommended
```

Comment:
What does it mean to have a segmented button with only one segment?  That seems very weird, UI-wise.

Reply:
We should restrict to 2-5

Reply:
Sorry for intruding here, I'm just a Flutter user and got a notification because I subscribed to this issue, so feel free to ignore this comment. Personally I would argue against artificially restricting the number of segments as long as an arbitrary number of segments doesn't inherently break anything with the representation of the button. I only see the following problems with a large number of segments:
1. It can easily overflow or have similar issues.
2. It's against the material design guidelines.

In my opinion these points don't justify having this hard restriction which would require a user with a potentially valid use case to duplicate the entire code of this 700 line widget (plus maybe the theme data). In particular:

1. Overflow or similar issues can also occur with e.g. five very long segments.
2. There are many places in the library where a user can do stuff that is against the material design guidelines (e.g. using wrong paddings, choosing colors that don't fit according to the spec etc.). It is not a clear case to me that Flutter should enforce the material guidelines. Maybe a user has a different design concept for the entire app but the general functionality of this segmented button is still useful for them. In any case the documentation could just link to the guidelines for the suggested maximum of 5 items. I think there are other widgets that also refer to the guidelines to _suggest_ a conforming design.

For me as a developer it would appear very arbitrary if there was e.g. an `assert(segments.length < 5)` statement in the code.

Reply:
Good points, thank you. I just noticed the recommendation below for two to five. It sounds reasonable to avoid specifying an upper limit, but a single segmented button doesn't make sense UI wise. A button or chip or checkbox or switch would be more appropriate in those cases.

Reply:
Yeah, I thought about restricting it to 5, but for the reasons that @w1th0utnam3 mentioned above I went with just the recommendation. 

As for allowing a single segment, while it would be a little weird, there is nothing technically wrong in terms of the implementation. If people feel strongly about this, we can add an assert that there are at least 2 segments.

Reply:
Maybe just add a comment mentioning that a one-segment segmented button is probably better represented as a choice chip, no need to assert that there is more than one.